PERT distribution and basic test for issue #209

colemakdvorak opened the pull request:
Hello,

Issue #209 looked like a good first issue to learn about some building blocks of this project so I gave it a go. I held off on KL divergence implementation for now since I couldn't see a clean analytical solution right off the bat.

This PR implements the [modified PERT](https://en.wikipedia.org/wiki/PERT_distribution), which is a general case of standard PERT that includes an extra parameter to control the shape of the distribution. 

I used TensorFlow pylint for general style guide while referencing other implementations within this repo at my own discretion. So it might be a little off here and there - please feel free to point fingers. Whether it's about style, documentation, or implementation, I am open to any suggestions.

brianwa84 commented on tensorflow_probability/python/distributions/pert.py:
```diff
@@ -0,0 +1,164 @@
+      dtype = dtype_util.common_dtype([mini, maxi, mode], tf.float32)
+      self._min = tensor_util.convert_nonref_to_tensor(
+          mini, name='mini', dtype=dtype)
+      self._mod = tensor_util.convert_nonref_to_tensor(
```
should this be `_mode`?

colemakdvorak replied:
I guess this won't matter anymore since I agree on changing these names to `low`, `peak`, and `high` but I am just curious about how to approach this kind of case in terms of general software engineering / design principle.

So I avoided calling this `_mode` because it seemed to be an abstract method in `distribution` class - and I didn't want to set a property on a name where a class usually expects a method. This is how I ended up setting it to `_mod` and implemented `_mode` method to return this property. Is this approach reasonable or is there a general good/alternative practice when it comes to these kind of things?

brianwa84 commented on tensorflow_probability/python/distributions/pert.py:
```diff
@@ -0,0 +1,164 @@
+]
+
+class PERT(transformed_distribution.TransformedDistribution):
+  """Modified PERT distribution:
```
the first line should be a one-liner "sentence" e.g. "Modified PERT distribution for modeling expert predictions."
then an empty line.

can you add a "Mathematical details" section a la https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/distributions/gamma_gamma.py#L51
and optionally an example or two a la
https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/distributions/dirichlet.py#L103 ?

colemakdvorak replied:
Added a simple mathematical overview and couple of examples demonstrating broadcasting behaviors.

https://github.com/tensorflow/probability/blob/33ea65ee4be4ed05247357f7d5b374a38d3b3f58/tensorflow_probability/python/distributions/pert.py#L21

brianwa84 commented on tensorflow_probability/python/distributions/pert.py:
```diff
@@ -0,0 +1,164 @@
+  For general reference: https://en.wikipedia.org/wiki/PERT_distribution
+  """
+  def __init__(self,
+               mini,
```
let's rename these to match the triangular distribution (low, high, peak [though mode is kind of nice too])
https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/distributions/triangular.py#L95
you might even mention PERT being "like Triangular, but smoother" or something along those lines.

colemakdvorak replied:
Done.

brianwa84 commented on tensorflow_probability/python/distributions/pert.py:
```diff
@@ -0,0 +1,164 @@
+          maxi, name='maxi', dtype=dtype)
+      self._smoothness = tensor_util.convert_nonref_to_tensor(
+          smoothness, name='smoothness', dtype=dtype)
+      self._concentration1 = 1. + self._smoothness \
```
for long lines, tfp style is to put an open close paren around the whole thing and omit the `\`

brianwa84 commented on tensorflow_probability/python/distributions/pert.py:
```diff
@@ -0,0 +1,164 @@
+          validate_args=validate_args,
+          parameters=parameters,
+          name=name)
+
```
can you also implement _batch_shape (which is preferred when shapes are statically known)

brianwa84 commented on tensorflow_probability/python/distributions/pert.py:
```diff
@@ -0,0 +1,164 @@
+    return assertions
+
+  def _survivial_function(self, value, **kwargs):
+    return 1. - self.cdf(value, kwargs)
```
this should already be implemented as such in the base class (i.e. you can leave it out)

brianwa84 commented on tensorflow_probability/python/distributions/pert.py:
```diff
@@ -0,0 +1,164 @@
+        prefer_static.shape(
+            self.concentration0 if concentration0 is None else concentration0))
+
+  def _parameter_control_dependencies(self, is_init):
```
we prefer sticking this method at the very bottom, out of the way :)

brianwa84 commented on tensorflow_probability/python/distributions/pert.py:
```diff
@@ -0,0 +1,164 @@
+        assertions.append(assert_util.assert_greater(
+            self._mod,
+            self._min,
+            message="Mode must be greater than minimum."
```
minor nit: can we put these param names in backticks like "`peak` must be greater than `low`"

brianwa84 commented on tensorflow_probability/python/distributions/pert_test.py:
```diff
@@ -0,0 +1,205 @@
+        mini=[1., 2.],
+        mode=[[2., 3.], [4., 5.]],
+        maxi=10.,
+        smoothness=4.,
```
I'm guessing we might want to rename this to `temperature`, but if you want you can leave it as is for now, I'll discuss w/ the team when we get ready to merge it.

colemakdvorak replied:
To be honest either works for me intuitively, but I've seen the term `temperature` several times in context of statistics so I changed it.

brianwa84 commented on tensorflow_probability/python/distributions/pert_test.py:
```diff
@@ -0,0 +1,205 @@
+      dist = tfd.PERT(mini, mode, maxi, validate_args=True)
+      self.evaluate(dist.sample(1))
+
+  def testMaxModeInequalityAfterMutation(self):
```
this test looks great!

if you want a little more coverage, you can add a 
```python
@classmethod
def _params_event_ndims(cls):
  return dict(low=0, high=0, peak=0, temperature=0)
```
to the distribution (which will also make it batch-sliceable, i.e. `tfd.PERT(...)[..., 3:, -2]`), and then whitelist it in distribution_properties_test.py

colemakdvorak replied:
Cool, I haven't fully wrapped my head around details of `distribution_properties_test.py` but I get the gist of it.

https://github.com/tensorflow/probability/blob/ba0701f00cd1bed083caa61f392f4acd91f0fe20/tensorflow_probability/python/distributions/distribution_properties_test.py#L1048

I saw the test inputting some wild parameters, so I fixed it by adding something similar to `fix_triangular` with additional handling of `temperature`.

brianwa84 replied:
Great, I'll double check it when I bring the code in. I wonder if rebasing the PR instead of adding the commits into this PR would fix the CLA issue. It seems to have a problem with the "no-reply@google.com" commit (I think, the PlackettLuce one).

colemakdvorak commented:
Oops. In the process of fixing the suggestions and keeping my forked repo up to date I ended up including commits that don't include CLA.

colemakdvorak commented:
Thank you for such an extensive feedback @brianwa84 . I went through all of your feedbacks and addressed the points you brought up.

Not sure what to do about the CLA part though, I assume it's got to do with me including a Googler's commit. Can I expect this to go away or do I juggle with git?

csuter commented:
are you merging from master or rebasing on your fork? using rebase would definitely leave a cleaner trail and avoid this issue. i generally keep a local branch called "upstream" which tracks the remote master, pull to that, then rebase my local master on top of it. i can help sort things out from this state if you need a hand.

csuter commented:
i'm working on creating a copy of your branch now without the merges; will update shortly. once it's in good shape you can just (force) pull it onto your master and the PR will automatically clean itself up.

csuter commented:
Ok, so I am a little confused about what happened before 20535835: "merge conflict resolution" (it looks like some commits contained unresolved merge conflict markers). So I've dropped everything before that since it looks like you tried to start over from scratch.

There's also been some annoyingness because PERT is being checked in simultaneously with PlackettLuce, and git is having a really hard time with all the edits on adjacent lines. Furthermore, PlackettLuce was submitted, then just now rolled back, and will shortly be resubmitted again, so we're probably not out of the woods with those merge conflicts 😭

Anyway, here is a branch on my fork that has everything you've done so far (aside from the stuff before 20535835: "merge conflict resolution", as noted above):

https://github.com/csuter/probability/tree/fix_colemakdvorak

Here's what you can do to clean up this PR:

```
# 1. First things first, create a branch that knows where your master is now,
# so you can get back to it later if need be.
git checkout master
git checkout -b master_backup

# 2. Add my fork as a remote, and fetch its contents
git remote add csuter https://github.com/csuter/probability
git fetch csuter

# 3. Hard reset your master to point to where my fixer upper branch is
git reset --hard csuter/fix_colemakdvorak

# 4. Update your remote on GitHub, which should result in this PR containing just your commits.
git push origin master
```

After that's all done your changes will be "rebased" onto the current HEAD of the TFP repo, the CLA  issues should disappear. Once Aditya rolls back the rollback of PluckettLuce, you'll probably hit some more merge issues. Maybe we can get your PR in before then. If not, I strongly advise using rebase locally to get things in order again. Assuming you have a local branch called `upstream` that tracks the TFP repo, this would look something like

```
# Switch to upstream branch and get the latest goodies
git checkout upstream
git pull

# Jump back to your master branch and rebase
git checkout master
git rebase upstream
```

If you hit conflicts there, git should give some nice instructions for resolving. Basically, edit the conflicting files (they'll be shown in red if you do `git status`), resolve the merge conflicts (marked by `<<<<<<<<`, `>>>>>>>`, `=======`), `git add` the fixed up files, then run `git rebase --continue` to proceed. Rebase is stepping through each of your commits to try to "replay" them on top of the upstream, so you may run into multiple of these, as say one of your commits collided with PlackettLuce in BUILD and another in __init__.py (or whatever).

Finally, this is a super useful command for visualizing the commit history in the form of a graph:

```
git -c color.diff=always log --graph --date-order --date=short --branches --pretty='%C(green bold)%h%Creset %C(blue bold)%d%Creset %C(black bold)%s%Creset
```

I have this aliased in my .gitconfig as `git gr`. When I'm doing git surgery, i basically alwasy keep an extra terminal open off to the side with the output of this command, and refresh it each time I make a change. I probably could not use git without this command :)

Please let me know if you have any questions -- sorry for the insanely long post!

colemakdvorak commented:
Sorry about the hassle here @csuter . I will start looking into your instructions. I've been fetching then rebasing from upstream/master to my local master, and at some point there has been a pull/merge conflict between my forked repo and local repo. Which I dealt with individually. And what I did is what you got :cry:.

csuter commented:
No hassle at all. For some reason I get a great deal of pleasure out of fixing git issues 🤷‍♂️

When you say "rebasing from upstream/master to local master" do you mean that the other way around? I think the jargon would have it "rebasing my local master onto the upstream master" which means "replaying some local commits, which were written on top of an old upstream state, as if they had been written on top of a newer upstream state". If that's what you meant, great! (Although then I'm not sure how the issues arose in the PR). If you meant something else, then that may be a source of the issues that arose here. If you run that graph command I shared on your repo in its current state, you'll see there are some branches that diverged and then rejoined. That implies there must've been some git merge commands run at some point (maybe in the GH UI?). Git merges are basically the root of all (VCS) evil IMO 😊. I wish GH didn't make them so easy to do with the click of a button....

colemakdvorak commented:
Well god bless, thanks for the git surgery instructions - the CLA sign is green now. And yes, I think I got terminology direction backwards. I will take note of the tricks and tips, and try to brush up on git. Thank you, it really means a lot to me.

csuter commented:
Woohoo! Glad we got things sorted out.

Looks like PlackettLuce is back in now, so there's a good chance to practice those git tricks, getting your master rebased onto upstream again :)

Feel free to reach back out with any questions or issues you run into!

colemakdvorak commented:
I checked few changes to commits done and I've got to say I learned more about this project than I contributed :) 

I know that the style guide already refers to calling `convert_to_tensor` on every argument after a name scope, and few implemented methods like `_mode` and `_variance` are called through a wrapper function that defines the scope and I missed that.

Quick question before closing:

If I am understanding `_transformed_beta` correctly, I assume invoking the method doesn't instantiate a transformed beta distribution every time it's called since local variables in that method are tied to specific properties defined at construction. Then what is the reasoning behind converting some arguments in `_variance` and passing it in as an argument again? What does "avoiding multiple conversions" in comment mean?

brianwa84 commented:
> learned more about this project than I contributed
I don't think so! Without your work here, there would be no `tfd.PERT`. I learned some things too, e.g. about the statistical testing thing in the empirical cdf test during internal review.

Most of the changes I made while pulling it in were to make it more compliant with a "GradientTape-safety" effort we're working through. The specific goal would be to be able to do something like:

```python
d = tfd.PERT(tf.Variable(0.), tf.Variable(1.), tf.Variable(.7))
d.low.assign(.1)  # don't miss this assignment!
with tf.GradientTape() as tape:
  nll = -d.log_prob(.5)
grads = tape.gradient(nll, d.trainable_variables)  # not None, please!
```

The problem is that if we create concentration0/1 in the constructor, we will lose the linkage to those Variables: if those variables change the distribution will not reflect the changes; if we ask for gradients to the variables, they will be `None`. So we do indeed reconstruct the underlying transformed beta for each relevant method call, to ensure that variables are respected.

While we are trying to respect these variables, we are also trying to make sure we don't create too many reads (which require synchronization, and could be racy with variables assignments). This is what "multiple conversions" is referring to (`tf.convert_to_tensor(a_variable)`). I also made more rigorous the `distribution_properties_test`, to ensure we catch a couple more cases that were slipping through e.g. `fix_triangular` and `fix_pert`.

Hopefully that gives you a bit of background on the diff from your PR to head. Mostly it was moving your code around, removing sqrt(variance) because it's implemented in tfd.Distribution base class, little things that depend on some background knowledge.

Thanks for your PR, and lmk if you  have more questions; I'm happy to answer. Please keep sending more PRs!! Distributions & bijectors are quite valuable to increasing the breadth of the library, and we're happy to keep enriching diagnostics like tfp.stats, tfp.math, tfp.math.psd_kernels, etc.
